worth2/main/generic/models.py

```diff
@@ -59,29 +59,31 @@ def last_access_in_module(self, module_num):
             '-last_visit').first()
 
     def last_location_url(self):
-        if self.percent_complete() == 0:
+        if self.percent_complete_hierarchy() == 0:
             return reverse('root')
         else:
             return self.last_access_hierarchy().section.get_absolute_url()
 
-    def percent_complete(self):
-        return self.percent_complete_hierarchy()
-
-    def percent_complete_hierarchy(self):
-        hierarchy = Hierarchy.get_hierarchy('main')
-        pages = hierarchy.get_root().get_descendants().count()
+    def percent_complete_by_pages(self, pages):
+        page_count = pages.count()
 
-        if pages > 0:
-            visits = UserPageVisit.objects.filter(user=self.user).count()
-            return int(visits / pages * 100)
+        if page_count > 0:
+            visits = UserPageVisit.objects.filter(
+                user=self.user, section__in=pages).count()
+            return int(visits / page_count * 100)
```

Comment:
I *think* we want to cast `visits` and/or `page_count` to a float in there to avoid surprising integer division problems in python2:

```
>>> visits = 3
>>> page_count = 10
>>> int(visits / page_count * 100)
0
```